source/common/network/dns_resolver/dns_factory.h

```diff
@@ -0,0 +1,165 @@
+    envoy::config::core::v3::TypedExtensionConfig& typed_dns_resolver_config) {
+  envoy::extensions::network::dns_resolver::cares::v3::CaresDnsResolverConfig cares;
+  cares.mutable_dns_resolver_options()->set_use_tcp_for_dns_lookups(
+      config.use_tcp_for_dns_lookups());
+  typed_dns_resolver_config.mutable_typed_config()->PackFrom(cares);
+  typed_dns_resolver_config.set_name(std::string(CaresDnsResolver));
+}
+
+// Overloading the template function for DnsFilterConfig type, which doesn't need to copy anything.
+void handleLegacyDnsResolverData(
+    const envoy::extensions::filters::udp::dns_filter::v3alpha::DnsFilterConfig::
+        ClientContextConfig&,
+    envoy::config::core::v3::TypedExtensionConfig& typed_dns_resolver_config);
+
+// Overloading the template function for Cluster config type, which need to copy
+// both use_tcp_for_dns_lookups and dns_resolvers.
+void handleLegacyDnsResolverData(
+    const envoy::config::cluster::v3::Cluster& config,
+    envoy::config::core::v3::TypedExtensionConfig& typed_dns_resolver_config);
+
+// Make typed_dns_resolver_config from the passed @param config.
+template <class ConfigType>
+envoy::config::core::v3::TypedExtensionConfig makeDnsResolverConfig(const ConfigType& config) {
+  envoy::config::core::v3::TypedExtensionConfig typed_dns_resolver_config;
+
+  // typed_dns_resolver_config takes precedence
+  if (checkTypedDnsResolverConfigExist(config, typed_dns_resolver_config)) {
+    return typed_dns_resolver_config;
+  }
+
+  // If use apple API for DNS lookups, create an AppleDnsResolverConfig typed config.
+  if (checkUseAppleApiForDnsLookups(typed_dns_resolver_config)) {
+    return typed_dns_resolver_config;
+  }
+
+  // If dns_resolution_config exits, create a CaresDnsResolverConfig typed config based on it.
+  if (checkDnsResolutionConfigExist(config, typed_dns_resolver_config)) {
+    return typed_dns_resolver_config;
+  }
+
+  // Handle legacy DNS resolver fields for backward compatibility.
+  // Different config type has different fields to copy.
+  handleLegacyDnsResolverData(config, typed_dns_resolver_config);
+  return typed_dns_resolver_config;
+}
+
+// Create the DNS resolver factory from the typed config. This is the underline
+// function which performs the registry lookup based on typed config.
+Network::DnsResolverFactory& createDnsResolverFactoryFromTypedConfig(
+    const envoy::config::core::v3::TypedExtensionConfig& typed_dns_resolver_config);
+
+// Create the default DNS resolver factory. apple for MacOS or c-ares for all others.
+// The default registry lookup will always succeed, thus no exception throwing.
+// This function can be called in main or worker threads.
+Network::DnsResolverFactory& createDefaultDnsResolverFactory(
+    envoy::config::core::v3::TypedExtensionConfig& typed_dns_resolver_config);
+
+// Create the DNS resolver factory from the proto config.
+// Retrieve the DNS related configurations in the passed in @param config, and store the
```

Comment:
Nit: not sure if `@param config` is any clearer here than `config parameter`. Why not just use standard Doxygen comment format?

Reply:
ACK